### DISCUSSION TITLE
Pass data to javascript file

### DISCUSSION THREAD

**Discussion by franciscobmacedo** [django-components/django-components#399] [Q&A] [answered]
Is there a way to pass template context data to the javascript file? 

**usecase**: I want to use a unique ID attribute for an html element and I want to query said element in the javascript file.

**template**
```html

<div id="{{unique_id}}">...</div>
```

**javascript**
```javascript

// I'm assuming the varialbe is passed along through 'data' attributes

const data = document.currentScript.dataset;
const unique_id = data.unique_id;
const element = document.querySelector(unique_id);
```

**Comment by EmilStenstrom** [comment]
Hi @franciscobmacedo! Great question. I'm thinking we should maybe create a unique ID for each instance of a component automatically and make that easy accessible. 

But for now, I think you could [override render_js_dependencies ](https://github.com/EmilStenstrom/django-components/blob/master/django_components/component.py#L121C9-L121C31) and just replace `<script` with `<script data...`. Does that work?

**Reply by EmilStenstrom** [reply, marked as answer]
Oh wait. If you specify your javascript as a dependency it will just be rendered once, which kind of defeats the purpose. I guess you could put the script tag in the template file?

**Reply by franciscobmacedo** [reply]
Thanks for the quick reply! Creating a unique id for each component is a great idea.

Yes, that solution works, if one is writing javascript directly in a python file.

However, I was trying to keep js code in a separate file and to pass this data to the specified js file only (so that I can safely use `document.currentScript` in the js file).

I belive in that case I would have to somehow override the django widgets [Media](https://github.com/django/django/blob/main/django/forms/widgets.py#L62) `render_js` function - which I'm not sure how to do it?

Or maybe you have another suggestion to achieve this?  I'm also happy to investigate

**Reply by franciscobmacedo** [reply]
Ah,ok, I didn't know that the dependent js file will only be rendered once even if there are multiple instances of the component in the same page - but that makes sense. 

Ok, in that case, I guess one has to keep the js code inside the template to make this work

**Reply by EmilStenstrom** [reply]
Could you define a function in the js dependency, and call it with a unique ID from the template?

**Reply by franciscobmacedo** [reply]
The problem is that the js dependency is not available when the template calls the function. To ilustrate:

**template**
```html
<div class="component-calendar">
    Hello from calendar component!
</div>
<script data-id="{{id}}">
    const data = document.currentScript.dataset;
    buildCalendar(data.id)
</script>
```

**javascript file**
```javascript
  (function () {
    const buildCalendar = (id) => {
      console.log("id is", id);
    };
  })();
```

This throws an exception:

```javascript
Uncaught ReferenceError: buildCalendar is not defined
```

**Reply by franciscobmacedo** [reply]
ok, never mind, made it work if:

- `{% component_js_dependencies %}` is added inside the html `<header>` tag
- I don't use `(function () {....` - just declared the function like:

```javascript
 const buildCalendar = (id) => {
      console.log("id is", id);
    };
```

**Reply by EmilStenstrom** [reply]
Excellent. 

I wonder what the ideal way would be? This is a bit convoluted. Any ideas?

**Reply by franciscobmacedo** [reply]
I don't think this is very straightforward - but I wonder if we can add an `initComponent` function to the js dependency, and inject this part of the code under the hood?

So for the user, it would be something like this in the python code:

```python
def get_js_context_data(self, param1, param2):
    return {
            "param1": param1,
            "param2": param2
    }
```

and in their js file:
```javascript
function initComponent(context){
    console.log(context)
    // ouptuts -> {'param1': 'param1',  'param2': 'param2'}
}
```

and under the hood, django-componentes can use the bult-in [json_script filter](https://docs.djangoproject.com/en/5.0/ref/templates/builtins/#json-script) and pass something like this:

```python
def get_template(self, context):
    template = ... # same logic as before
    js_context = self.get_js_context_data(...)

    # parse context data to js
    data_script = f"{{ {js_context}|json_script:{self.component_id} }}" 

    # call initComponent
    data_script_loader = f"""
       <script>
             const context = JSON.parse(document.getElementById({self.component_id}).textContent); 
             initComponent(context);
        </script>
     """ 
    return template + data_script + data_script_loader
```

But I'm not sure, it still feels convoluted and a bit of a black box. What do you think?

**Reply by ekaj2** [reply]
This is where I've landed on for now:

In the template:

```html
<div class="component-my_component">

    <!-- rest of component ... -->

    <script>
        (function() {  // don't pollute window w/ this const
            const component = document.currentScript.parentElement;
            window.addEventListener('load', function() {
                initializeMyComponent(component);
            });
        })()
    </script>
</div>
```

In script.js:

```js
function initializeMyComponent(component) {
    console.log(`Loaded: ${component}`);
}
```

This delays the load long enough that initializeMyComponent is defined.

All of it is boilerplate though, the only component-specific piece would be naming the init function differently so it doesn't clash on the window namespace.

**Reply by JuroOravec** [reply]
@ekaj2 There's also a discussion going on in https://github.com/EmilStenstrom/django-components/issues/622. Good idea with wrapping the execution with `window.addEventListener('load', ...)`.

**Comment by JuroOravec** [comment]
Not an answer to this question directly, but want to throw it out there. In my project we use AlpineJS, and I structure django components to have (mostly) the same boundaries as AlpineJS components. And it works fairly well, because AlpineJS is in charge of ensuring that each AlpineJS component has its own JS context and is initialized when the component is rendered in the browser.

And to pass different data for different component instances, I pass the data from Django to HTML (as [dataset](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes)), and then from within AlpineJS I query the HTML element for the data.

AlpineJS also handles the case of generating unique IDs for their components on the spot. See https://alpinejs.dev/magics/id
